doc/specs/drafts/#5916 - Triggers/#5916 - Triggers.md

```diff
@@ -0,0 +1,601 @@
+            "pattern": "    git push --set-upstream origin ([^\\w]*)",
+            "runOn": "mark"
+        },
+        "action": "sendInput",
+        "input": "git push --set-upstream origin ${match[1]}\r"
+    },
+    {
+        "match": "'(.*)' is not recognized as an internal or external command,",
+        "action": "checkWinGet",
+        "commandline": "${match[1]}"
+    },
+    {
+        "match": "This will open a new Terminal tab",
+        "action": "newTab"
+    },
+]
+```
+
+`triggers` is an array of `Trigger` objects, which are a combination of an
+`ActionAndArgs`, a `Control.Matcher`, and a `Control.TriggerAction`.
+
+> **Note**
+> Interface and other types here are shown in C#, since that's fairly close to
+> MIDL3, which would be used for cross-component interfaces.
+
+```c#
+runtimeclass Terminal.Settings.Model.Trigger
+{
+    String Id;
+
+    Control.Matcher Matcher;
+    Control.TriggerAction ControlAction;
+
+    Terminal.Settings.Model.ActionAndArgs ParseMatches(String[] matches);
+
+    private String _actionJson;
+}
+```
+
+### Trigger matchers
+
+The `match` property is parsed as a `Control.Matcher`. This includes the
+following properties, which are heavily inspired by the matchers in VsCode.
+
+```jsonc
+"match": {
+    "anchor": "bottom",
+    "offset": 0, // starting how many rows from the bottom
+    "length": 5, // how many lines to match against
+    "pattern": "    git push --set-upstream origin ([^\\w]*)",
+    "runOn": "everything|newline|mark",
+    "buffer": "main|alt|any"
+}
+```
+
+* `"anchor"` (_default: `"bottom"`_) : Which part of the viewport to run the match against.
+  - VsCode has this, but I don't think we really do. I think we can just always say bottom.
+* `"offset"` (_default: `0`_) : Run the match starting how many rows from the anchor
+* `"length"` (_default: `1`_) : How many rows to include in the match
+* `"pattern"` (_required_) : The regex to run. If omitted / empty, this entire
+  trigger is ignored.
+* `"runOn"` (_default: `"newline"`_) : When to run this match.
+  * `"everything"`: On every print from the connection. High performance impact.
+  * `"newline"`: On every newline character from the connection. Medium
+    performance impact, depending on the workload.
+  * `"mark"`: Whenever a mark ([#11000]) is emitted to the terminal. With
+    `autoMarkPrompts`, this includes when the user presses <kbd>enter</kbd>.
+    This has a much smaller performance impact, assuming the user isn't marking
+    every line.
+* `"buffer"` (_default_: `"any"`): Configure which terminal buffer this trigger runs on.
+  * `main`: only run in the main buffer (e.g., most shells)
+  * `alt`: only run in the alternate screen buffer (e.g., `vim`, `tmux`, any full-screen application)
+  * `any`: Run in either buffer.
+
+`match` also accepts just a single string, to automatically assume default
+values for the other properties. For example, the JSON `"match":
+"[Ee][Rr][Rr][Oo][Rr]"` is evaluated as

```

Comment:
But why do that now? We don't even know yet how the feature will actually work in practice and thus don't really know if such shortcuts would be helpful either. Less ways to specify settings seems generally beneficial to me.

Reply:
My gut says that more often than not, I just want to match on a line of text as it is output, so I usually don't care about the rest of the match params. I wanted to make it easier for folks to write that into their settings, for folks who are editing by hand.

Reply:
I still think it'd be better to have just one way to specify settings, at least initially. But I didn't mean to imply that I'm against it either.